[HttpKernel][FrameworkBundle] Service definition of RequestPayloadValueResolver fails on linting without Serializer component

### Symfony version(s) affected

6.3

### Description

When updating [an application](https://github.com/chr-hertel/oss-complexity-report) to Symfony 6.3 beta, without serializer component installed, the `lint:container` command throws following exception:

```
 [ERROR] Invalid definition for service "argument_resolver.request_payload": argument 1 of                              
         "Symfony\Component\HttpKernel\Controller\ArgumentResolver\RequestPayloadValueResolver::__construct()" accepts  
         "Symfony\Component\Serializer\SerializerInterface&Symfony\Component\Serializer\Normalizer\DenormalizerInterface
         ", "Symfony\Component\Serializer\Serializer" passed.  
```

### How to reproduce

See https://github.com/chr-hertel/oss-complexity-report/pull/5

### Possible Solution

I guess the solution is somewhere around the service definition of RequestPayloadValueResolver

### Additional Context

Symfony version of `event-dispatcher` and `process` are on v5.4 in this case, but I don't think that this really matters here.

--- a/src/Symfony/Bundle/FrameworkBundle/Command/ContainerLintCommand.php
+++ b/src/Symfony/Bundle/FrameworkBundle/Command/ContainerLintCommand.php
@@ -88,8 +88,6 @@ final class ContainerLintCommand extends Command
                 return $this->buildContainer();
             }, $kernel, $kernel::class);
             $container = $buildContainer();
-
-            $skippedIds = [];
         } else {
             if (!$kernelContainer instanceof Container) {
                 throw new RuntimeException(sprintf('This command does not support the application container: "%s" does not extend "%s".', get_debug_type($kernelContainer), Container::class));
@@ -100,13 +98,6 @@ final class ContainerLintCommand extends Command
             $refl = new \ReflectionProperty($parameterBag, 'resolved');
             $refl->setValue($parameterBag, true);
 
-            $skippedIds = [];
-            foreach ($container->getServiceIds() as $serviceId) {
-                if (str_starts_with($serviceId, '.errored.')) {
-                    $skippedIds[$serviceId] = true;
-                }
-            }
-
             $container->getCompilerPassConfig()->setBeforeOptimizationPasses([]);
             $container->getCompilerPassConfig()->setOptimizationPasses([]);
             $container->getCompilerPassConfig()->setBeforeRemovingPasses([]);
@@ -115,7 +106,7 @@ final class ContainerLintCommand extends Command
         $container->setParameter('container.build_hash', 'lint_container');
         $container->setParameter('container.build_id', 'lint_container');
 
-        $container->addCompilerPass(new CheckTypeDeclarationsPass(true, $skippedIds), PassConfig::TYPE_AFTER_REMOVING, -100);
+        $container->addCompilerPass(new CheckTypeDeclarationsPass(true), PassConfig::TYPE_AFTER_REMOVING, -100);
 
         return $this->containerBuilder = $container;
     }

--- a/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
+++ b/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
@@ -375,13 +375,12 @@ class FrameworkExtension extends Extension
 
             $this->registerSerializerConfiguration($config['serializer'], $container, $loader);
         } else {
-            $container->register('.argument_resolver.request_payload.no_serializer', Serializer::class)
+            $container->getDefinition('argument_resolver.request_payload')
+                ->setArguments([])
                 ->addError('You can neither use "#[MapRequestPayload]" nor "#[MapQueryString]" since the Serializer component is not '
                     .(class_exists(Serializer::class) ? 'enabled. Try setting "framework.serializer" to true.' : 'installed. Try running "composer require symfony/serializer-pack".')
-                );
-
-            $container->getDefinition('argument_resolver.request_payload')
-                ->replaceArgument(0, new Reference('.argument_resolver.request_payload.no_serializer', ContainerInterface::RUNTIME_EXCEPTION_ON_INVALID_REFERENCE))
+                )
+                ->addTag('container.error')
                 ->clearTag('kernel.event_subscriber');
 
             $container->removeDefinition('console.command.serializer_debug');

--- a/src/Symfony/Bundle/FrameworkBundle/composer.json
+++ b/src/Symfony/Bundle/FrameworkBundle/composer.json
@@ -21,7 +21,7 @@
         "ext-xml": "*",
         "symfony/cache": "^5.4|^6.0",
         "symfony/config": "^6.1",
-        "symfony/dependency-injection": "^6.2.8",
+        "symfony/dependency-injection": "^6.3",
         "symfony/deprecation-contracts": "^2.5|^3",
         "symfony/error-handler": "^6.1",
         "symfony/event-dispatcher": "^5.4|^6.0",

--- a/src/Symfony/Component/DependencyInjection/Compiler/DefinitionErrorExceptionPass.php
+++ b/src/Symfony/Component/DependencyInjection/Compiler/DefinitionErrorExceptionPass.php
@@ -25,7 +25,7 @@ class DefinitionErrorExceptionPass extends AbstractRecursivePass
 {
     protected function processValue(mixed $value, bool $isRoot = false): mixed
     {
-        if (!$value instanceof Definition || !$value->hasErrors()) {
+        if (!$value instanceof Definition || !$value->hasErrors() || $value->hasTag('container.error')) {
             return parent::processValue($value, $isRoot);
         }
 

--- a/src/Symfony/Component/DependencyInjection/Dumper/XmlDumper.php
+++ b/src/Symfony/Component/DependencyInjection/Dumper/XmlDumper.php
@@ -129,7 +129,9 @@ class XmlDumper extends Dumper
             }
         }
 
-        foreach ($definition->getTags() as $name => $tags) {
+        $tags = $definition->getTags();
+        $tags['container.error'] = array_map(fn ($e) => ['message' => $e], $definition->getErrors());
+        foreach ($tags as $name => $tags) {
             foreach ($tags as $attributes) {
                 $tag = $this->document->createElement('tag');
                 if (!\array_key_exists('name', $attributes)) {

--- a/src/Symfony/Component/DependencyInjection/Dumper/YamlDumper.php
+++ b/src/Symfony/Component/DependencyInjection/Dumper/YamlDumper.php
@@ -69,7 +69,9 @@ class YamlDumper extends Dumper
         }
 
         $tagsCode = '';
-        foreach ($definition->getTags() as $name => $tags) {
+        $tags = $definition->getTags();
+        $tags['container.error'] = array_map(fn ($e) => ['message' => $e], $definition->getErrors());
+        foreach ($tags as $name => $tags) {
             foreach ($tags as $attributes) {
                 $att = [];
                 foreach ($attributes as $key => $value) {

--- a/src/Symfony/Component/DependencyInjection/Loader/FileLoader.php
+++ b/src/Symfony/Component/DependencyInjection/Loader/FileLoader.php
@@ -214,6 +214,12 @@ abstract class FileLoader extends BaseFileLoader
     {
         $this->container->removeBindings($id);
 
+        foreach ($definition->getTag('container.error') as $error) {
+            if (isset($error['message'])) {
+                $definition->addError($error['message']);
+            }
+        }
+
         if ($this->isLoadingInstanceof) {
             if (!$definition instanceof ChildDefinition) {
                 throw new InvalidArgumentException(sprintf('Invalid type definition "%s": ChildDefinition expected, "%s" given.', $id, get_debug_type($definition)));

--- a/src/Symfony/Component/DependencyInjection/Tests/Compiler/DefinitionErrorExceptionPassTest.php
+++ b/src/Symfony/Component/DependencyInjection/Tests/Compiler/DefinitionErrorExceptionPassTest.php
@@ -49,4 +49,18 @@ class DefinitionErrorExceptionPassTest extends TestCase
         $pass->process($container);
         $this->assertSame($def, $container->getDefinition('foo_service_id')->getArgument(0));
     }
+
+    public function testSkipErrorFromTag()
+    {
+        $container = new ContainerBuilder();
+        $def = new Definition();
+        $def->addError('Things went wrong!');
+        $def->addTag('container.error');
+        $container->register('foo_service_id')
+            ->setArguments([$def]);
+
+        $pass = new DefinitionErrorExceptionPass();
+        $pass->process($container);
+        $this->assertSame($def, $container->getDefinition('foo_service_id')->getArgument(0));
+    }
 }

--- a/src/Symfony/Component/DependencyInjection/Tests/Fixtures/xml/services9.xml
+++ b/src/Symfony/Component/DependencyInjection/Tests/Fixtures/xml/services9.xml
@@ -145,7 +145,9 @@
     <service id="runtime_error" class="stdClass" public="true">
       <argument type="service" id="errored_definition"/>
     </service>
-    <service id="errored_definition" class="stdClass"/>
+    <service id="errored_definition" class="stdClass">
+      <tag name="container.error" message="Service &quot;errored_definition&quot; is broken."/>
+    </service>
     <service id="preload_sidekick" class="stdClass" public="true">
       <tag name="container.preload" class="Some\Sidekick1"/>
       <tag name="container.preload" class="Some\Sidekick2"/>

--- a/src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/services9.yml
+++ b/src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/services9.yml
@@ -173,6 +173,8 @@ services:
         public: true
     errored_definition:
         class: stdClass
+        tags:
+            - container.error: { message: 'Service "errored_definition" is broken.' }
     preload_sidekick:
         class: stdClass
         tags:

